poc_iot_injector/src/receipt_txn.rs

```diff
@@ -0,0 +1,244 @@
+    Ok(poc_witnesses)
+}
+
+fn hz_to_mhz(freq_hz: u64) -> f32 {
+    let freq_mhz = Decimal::from(freq_hz) / Decimal::from(MHZTOHZ);
+    freq_mhz.to_f32().unwrap_or_default()
+}
+
+fn construct_poc_receipt(
+    beacon_report: Option<LoraValidBeaconReportV1>,
+    num_witnesses: u32,
+) -> Result<Option<BlockchainPocReceiptV1>> {
+    if let Some(beacon_report) = beacon_report {
+        let LoraValidBeaconReportV1 {
+            received_timestamp: _beacon_received_ts,
+            location: _beacon_location,
+            hex_scale: beacon_hex_scale,
+            report: beacon_report_req,
+        } = beacon_report;
+
+        if let Some(beacon_report_req) = beacon_report_req {
+            let LoraBeaconReportReqV1 {
+                pub_key: beacon_pub_key,
+                local_entropy: _beacon_local_entropy,
+                remote_entropy: _beacon_remote_entropy,
+                data: beacon_data,
+                frequency: beacon_frequency,
+                channel: beacon_channel,
+                datarate: beacon_datarate,
+                tx_power: beacon_tx_power,
+                timestamp: beacon_timestamp,
+                signature: beacon_signature,
+            } = beacon_report_req;
+
+            let beacon_hex_scale =
+                Decimal::from_f32_retain(beacon_hex_scale).unwrap_or_else(|| dec!(1.0));
+            let reward_unit = poc_challengee_reward_unit(num_witnesses)?;
+            let reward_shares = beacon_hex_scale * reward_unit;
+            let reward_shares = reward_shares.to_u32().unwrap_or_default();
+
+            // NOTE: signal, origin, snr and addr_hash are irrelevant now
+            // TODO: Attach reward_share using beacon_hex_scale
+            Ok(Some(BlockchainPocReceiptV1 {
+                gateway: beacon_pub_key,
+                timestamp: beacon_timestamp,
+                signal: 0,
+                data: beacon_data,
+                origin: 0,
+                signature: beacon_signature,
+                snr: 0.0,
+                frequency: hz_to_mhz(beacon_frequency),
+                channel: beacon_channel,
+                datarate: beacon_datarate.to_string(),
+                tx_power: beacon_tx_power,
+                addr_hash: vec![],
+                reward_shares,
+            }))
+        } else {
+            Ok(None)
+        }
+    } else {
+        Ok(None)
+    }
+}
+
+fn hash_txn(txn: &BlockchainTxnPocReceiptsV2) -> (Vec<u8>, String) {
+    let mut txn = txn.clone();
+    txn.signature = vec![];
+    let digest = Sha256::digest(&txn.encode_to_vec()).to_vec();
+    (
+        digest.clone(),
+        base64::encode_config(&digest, base64::URL_SAFE_NO_PAD),
+    )
+}
+
+fn sign_txn(txn: &BlockchainTxnPocReceiptsV2, keypair: &Keypair) -> Result<Vec<u8>> {
+    let mut txn = txn.clone();
+    txn.signature = vec![];
+    Ok(keypair.sign(&txn.encode_to_vec())?)
+}
+
+fn poc_challengee_reward_unit(num_witnesses: u32) -> Result<Decimal> {
+    if num_witnesses == 0 {
```

Comment:
shouldn't the verifier be marking any poc reports with 0 witnesses as invalid already?

Reply:
This should be addressed in a subsequent PR.. Can we add a note here?

Reply:
Added a todo 13c8771